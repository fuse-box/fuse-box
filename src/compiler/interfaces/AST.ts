import { IASTScope } from '../Visitor/Visitor';

export interface ASTNode {
  raw?: string;
  type: keyof typeof ASTType;

  $assign_pattern?: boolean;
  $fuse_classInitializers?: Array<ASTNode>;
  $fuse_class_declaration_visited?: boolean;
  $fuse_decoratorForClassIdentifier?: string;
  $fuse_decoratorsCaptured?: boolean;
  $fuse_decoratorsProcessed?: boolean;
  $fuse_visited?: boolean;
  $transpiled?: boolean;
  alternate?: ASTNode;
  base?: ASTNode;
  chain?: ASTNode;
  consequent?: ASTNode;
  decorators?: Array<ASTNode>;
  definite?: boolean;
  elementTypes?: Array<ASTNode>;
  elements?: Array<ASTNode>;
  expressions?: Array<ASTNode>;
  initializer?: ASTNode;
  literal?: ASTNode;
  moduleReference?: ASTNode;
  optional?: boolean;
  prefix?: boolean;
  returnType?: ASTNode;
  scope?: IASTScope;
  shorthand?: boolean;
  source?: ASTNode;
  sourceType?: string;
  specifiers?: Array<ASTNode>;
  test?: ASTNode;
  typeAnnotation?: ASTNode;
  typeName?: ASTNode;
  loc?: { end: { column: number; line: number }; start: { column: number; line: number } };

  imported?: ASTNode;
  members?: Array<ASTNode>;
  method?: boolean;

  JSXElement?: ASTNode;

  abstract?: boolean;
  accessibility?: string;
  argument?: ASTNode;
  arguments?: Array<ASTNode>;
  async?: boolean;
  attributes?: Array<ASTNode>;
  body?: ASTNode | Array<ASTNode>;
  callee?: ASTNode;
  children?: Array<ASTNode>;
  closingElement?: ASTNode;
  computed?: boolean;
  declaration?: ASTNode;
  declarations?: Array<ASTNode>;
  declare?: boolean;
  exported?: ASTNode;
  expression?: any;
  generator?: boolean;
  id?: ASTNode;
  implements?: Array<ASTNode>;
  init?: null | ASTNode;
  key?: ASTNode;
  kind?: string;
  left?: ASTNode;
  local?: ASTNode;
  name?: any;
  object?: ASTNode;
  openingElement?: ASTNode;
  operator?: string;
  parameter?: ASTNode;
  params?: Array<ASTNode>;
  properties?: Array<ASTNode>;
  property?: ASTNode;
  quasi?: ASTNode;
  quasis?: Array<ASTNode>;
  right?: ASTNode;
  static?: boolean;
  superClass?: ASTNode;
  tag?: ASTNode;
  tail?: ASTNode;
  value?: any;
}

export const ASTType = {
  EnumDeclaration: 'TSEnumDeclaration',
  EnumMember: 'TSEnumMember',

  ParameterProperty: 'TSParameterProperty',
  StringKeyword: 'TSStringKeyword',
  TypeAnnotation: 'TSTypeAnnotation',

  AbstractClassProperty: 'TSAbstractClassProperty',
  AbstractMethodDefinition: 'TSAbstractMethodDefinition',
  AsExpression: 'TSAsExpression',
  DeclareFunction: 'TSDeclareFunction',
  DeclareKeyword: 'TSDeclareKeyword',
  InterfaceDeclaration: 'TSInterfaceDeclaration',
  NonNullExpression: 'TSNonNullExpression',
  TypeAliasDeclaration: 'TSTypeAliasDeclaration',
  TypeAssertion: 'TSTypeAssertion',

  ArrayExpression: 'ArrayExpression',
  ArrayPattern: 'ArrayPattern',
  ArrowFunctionExpression: 'ArrowFunctionExpression',
  AssignmentExpression: 'AssignmentExpression',
  AssignmentPattern: 'AssignmentPattern',
  AwaitExpression: 'AwaitExpression',
  BigIntLiteral: 'BigIntLiteral',
  BinaryExpression: 'BinaryExpression',
  Block: 'Block',
  BlockStatement: 'BlockStatement',
  BreakStatement: 'BreakStatement',
  CallExpression: 'CallExpression',
  CatchClause: 'CatchClause',
  ClassBody: 'ClassBody',
  ClassDeclaration: 'ClassDeclaration',
  ClassExpression: 'ClassExpression',
  ClassProperty: 'ClassProperty',
  ConditionalExpression: 'ConditionalExpression',
  ContinueStatement: 'ContinueStatement',
  DebuggerStatement: 'DebuggerStatement',
  Decorator: 'Decorator',
  DoWhileStatement: 'DoWhileStatement',
  EmptyStatement: 'EmptyStatement',
  ExportAllDeclaration: 'ExportAllDeclaration',
  ExportDefaultDeclaration: 'ExportDefaultDeclaration',
  ExportNamedDeclaration: 'ExportNamedDeclaration',
  ExportSpecifier: 'ExportSpecifier',
  ExpressionStatement: 'ExpressionStatement',
  ForInStatement: 'ForInStatement',
  ForOfStatement: 'ForOfStatement',
  ForStatement: 'ForStatement',
  FunctionDeclaration: 'FunctionDeclaration',
  FunctionExpression: 'FunctionExpression',
  HTMLClose: 'HTMLClose',
  HTMLOpen: 'HTMLOpen',
  Identifier: 'Identifier',
  IfStatement: 'IfStatement',
  Import: 'Import',
  ImportDeclaration: 'ImportDeclaration',
  ImportDefaultSpecifier: 'ImportDefaultSpecifier',
  ImportNamespaceSpecifier: 'ImportNamespaceSpecifier',
  ImportSpecifier: 'ImportSpecifier',
  JSXAttribute: 'JSXAttribute',
  JSXClosingElement: 'JSXClosingElement',
  JSXClosingFragment: 'JSXClosingFragment',
  JSXElement: 'JSXElement',
  JSXEmptyExpression: 'JSXEmptyExpression',
  JSXExpressionContainer: 'JSXExpressionContainer',
  JSXFragment: 'JSXFragment',
  JSXIdentifier: 'JSXIdentifier',
  JSXMemberExpression: 'JSXMemberExpression',
  JSXOpeningElement: 'JSXOpeningElement',
  JSXOpeningFragment: 'JSXOpeningFragment',
  JSXSpreadAttribute: 'JSXSpreadAttribute',
  JSXSpreadChild: 'JSXSpreadChild',
  JSXText: 'JSXText',
  LabeledStatement: 'LabeledStatement',
  Line: 'Line',
  Literal: 'Literal',
  LogicalExpression: 'LogicalExpression',
  MemberExpression: 'MemberExpression',
  MetaProperty: 'MetaProperty',
  MethodDefinition: 'MethodDefinition',
  NewExpression: 'NewExpression',
  ObjectExpression: 'ObjectExpression',
  ObjectPattern: 'ObjectPattern',
  OptionalCallExpression: 'OptionalCallExpression',
  OptionalMemberExpression: 'OptionalMemberExpression',
  Program: 'Program',
  Property: 'Property',
  RestElement: 'RestElement',
  ReturnStatement: 'ReturnStatement',
  SequenceExpression: 'SequenceExpression',
  SpreadElement: 'SpreadElement',
  Super: 'Super',
  SwitchCase: 'SwitchCase',
  SwitchStatement: 'SwitchStatement',
  TaggedTemplateExpression: 'TaggedTemplateExpression',
  TemplateElement: 'TemplateElement',
  TemplateLiteral: 'TemplateLiteral',
  ThisExpression: 'ThisExpression',
  ThrowStatement: 'ThrowStatement',
  TryStatement: 'TryStatement',
  module: 'module',
  script: 'script',

  AbstractKeyword: 'AbstractKeyword',

  AnyKeyword: 'TSAnyKeyword',
  ArrayType: 'TSArrayType',

  AsyncKeyword: 'AsyncKeyword',
  BigIntKeyword: 'BigIntKeyword',
  BooleanKeyword: 'TSBooleanKeyword',
  CallSignatureDeclaration: 'CallSignatureDeclaration',
  ClassImplements: 'ClassImplements',
  ConditionalType: 'ConditionalType',
  ConstructSignatureDeclaration: 'ConstructSignatureDeclaration',
  ConstructorType: 'ConstructorType',
  EmptyBodyFunctionExpression: 'EmptyBodyFunctionExpression',
  ExportAssignment: 'ExportAssignment',

  ExportKeyword: 'ExportKeyword',
  ExternalModuleReference: 'ExternalModuleReference',
  FunctionType: 'TSFunctionType',
  ImportEqualsDeclaration: 'TSImportEqualsDeclaration',
  ImportExpression: 'ImportExpression',
  ImportType: 'ImportType',
  IndexSignature: 'IndexSignature',
  IndexedAccessType: 'IndexedAccessType',
  InferType: 'InferType',

  InterfaceBody: 'InterfaceBody',
  InterfaceHeritage: 'InterfaceHeritage',
  IntersectionType: 'IntersectionType',
  LiteralType: 'TSLiteralType',
  MappedType: 'MappedType',
  MethodSignature: 'MethodSignature',
  ModuleBlock: 'ModuleBlock',
  ModuleDeclaration: 'TSModuleDeclaration',
  NamespaceExportDeclaration: 'NamespaceExportDeclaration',
  NeverKeyword: 'TSNeverKeyword',

  NullKeyword: 'TSNullKeyword',
  NumberKeyword: 'TSNumberKeyword',
  ObjectKeyword: 'ObjectKeyword',
  OptionalType: 'OptionalType',

  ParenthesizedType: 'ParenthesizedType',
  PrivateKeyword: 'PrivateKeyword',
  PropertySignature: 'PropertySignature',
  ProtectedKeyword: 'ProtectedKeyword',
  PublicKeyword: 'PublicKeyword',
  QualifiedName: 'QualifiedName',
  ReadonlyKeyword: 'ReadonlyKeyword',
  RestType: 'RestType',
  StaticKeyword: 'StaticKeyword',

  SymbolKeyword: 'SymbolKeyword',
  ThisType: 'ThisType',
  TupleType: 'TSTupleType',

  JSDocNullableType: 'JSDocNullableType',
  JSDocUnknownType: 'JSDocUnknownType',
  TypeLiteral: 'TypeLiteral',
  TypeOperator: 'TypeOperator',
  TypeParameter: 'TypeParameter',
  TypeParameterDeclaration: 'TypeParameterDeclaration',
  TypeParameterInstantiation: 'TypeParameterInstantiation',
  TypePredicate: 'TypePredicate',
  TypeQuery: 'TypeQuery',
  TypeReference: 'TSTypeReference',
  UnaryExpression: 'UnaryExpression',
  UndefinedKeyword: 'TSUndefinedKeyword',
  UnionType: 'UnionType',
  UnknownKeyword: 'UnknownKeyword',
  UpdateExpression: 'UpdateExpression',
  VariableDeclaration: 'VariableDeclaration',
  VariableDeclarator: 'VariableDeclarator',
  VoidKeyword: 'TSVoidKeyword',
  WhileStatement: 'WhileStatement',
  WithStatement: 'WithStatement',
  YieldExpression: 'YieldExpression',
};
