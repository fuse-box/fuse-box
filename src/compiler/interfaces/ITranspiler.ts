import { ITarget } from '../../config/PrivateConfig';
export interface ITranspiler {
  ast: any;
  target: ITarget;
  moduleFileName?: string;
  env: { [key: string]: string };
  isBrowser: boolean;
  isServer: boolean;
}
