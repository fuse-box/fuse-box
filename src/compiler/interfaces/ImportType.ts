export enum ImportType {
  REQUIRE = 1,
  FROM = 2,
  RAW_IMPORT = 3,
  DYNAMIC = 4,
}
