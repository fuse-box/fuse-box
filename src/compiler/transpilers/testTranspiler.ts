import * as fs from 'fs';
import * as path from 'path';
import { ITarget } from '../../config/PrivateConfig';
import { createContext } from '../../core/Context';
import { createModule } from '../../core/Module';
import { createPackage } from '../../core/Package';
import { makeFuseBoxPath } from '../../utils/utils';
import { generate } from '../generator/generator';
import { ASTNode } from '../interfaces/AST';
import { parseTypeScript, parseJavascript } from '../parser';
import { transformCommonVisitors } from '../transformer';

export interface ICompileModuleProps {
  target?: ITarget;
  env?: { [key: string]: string };
  fileName?: string;
  code?: string;
  emitDecoratorMetadata?: boolean;
  useMeriyah?: boolean;
}

export function testTranspile(props: ICompileModuleProps) {
  let contents = props.code;

  if (!contents) {
    contents = fs.readFileSync(props.fileName).toString();
  }

  const ext = props.fileName ? path.extname(props.fileName) : '.tsx';

  let ast;
  if (props.useMeriyah) {
    ast = parseJavascript(contents, { jsx: true });
  } else {
    ast = parseTypeScript(contents, { jsx: true });
  }

  const ctx = createContext({
    cache: false,
    devServer: false,
    target: props.target || 'browser',
    env: props.env,
  });

  if (props.emitDecoratorMetadata) {
    ctx.tsConfig.compilerOptions.emitDecoratorMetadata = props.emitDecoratorMetadata;
  }

  const pkg = createPackage({ ctx: ctx, meta: {} as any });
  const module = createModule(
    {
      absPath: __filename,
      ctx: ctx,
      extension: ext,
      fuseBoxPath: makeFuseBoxPath(path.dirname(props.fileName || __filename), props.fileName || __filename),
    },
    pkg,
  );
  module.ast = ast as ASTNode;

  transformCommonVisitors(module);

  const res = generate(ast, {});

  return res;
}
