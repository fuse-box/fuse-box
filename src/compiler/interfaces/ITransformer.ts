import { ICompilerOptions } from '../../compilerOptions/interfaces';
import { Context } from '../../core/context';
import { IModule } from '../../moduleResolver/module';
import { IProductionContext } from '../../production/ProductionContext';
import { IVisit, IVisitorMod } from '../Visitor/Visitor';
import { ASTNode } from './AST';
import { ImportType } from './ImportType';

export type ITranformerCallback = (visit: IVisit) => undefined | IVisitorMod | void;

export interface ITransformerVisitors {
  onEachNode?: ITranformerCallback;
  onTopLevelTraverse?: ITranformerCallback;
}

export interface IRequireStatementModuleOptions {
  breakDependantsCache?: boolean;
}
export interface ITransformerCommon {
  compilerOptions: ICompilerOptions;
  ctx: Context;
  module: IModule;
  onRequireCallExpression?: (
    importType: ImportType,
    node: ASTNode,
    moduleOptions?: IRequireStatementModuleOptions,
  ) => void;
}

export interface ITransformerProduction {
  ctx: Context;
  module: IModule;
  productionContext: IProductionContext;
}

export interface ITransformer {
  target?: {
    test?: RegExp;
    type?: 'js' | 'js_ts' | 'ts';
  };
  commonVisitors?: (props?: ITransformerCommon) => ITransformerVisitors;
  productionWarmupPhase?: (props?: ITransformerProduction) => ITransformerVisitors;
}
