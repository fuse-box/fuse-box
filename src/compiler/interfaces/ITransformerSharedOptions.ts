import { ImportType } from './ImportType';
import { ASTNode } from './AST';

export interface ITransformerSharedOptions {
  onRequireCallExpression?: (importType: ImportType, node: ASTNode) => void;
}
