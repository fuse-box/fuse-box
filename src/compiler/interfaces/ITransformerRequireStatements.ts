import { ASTNode } from './AST';
import { IRequireStatementModuleOptions } from './ITransformer';
import { ImportType } from './ImportType';

export interface ITransformerRequireStatement {
  importType: ImportType;
  moduleOptions?: IRequireStatementModuleOptions;
  statement: ASTNode;
  value?: string;
}
export type ITransformerRequireStatementCollection = Array<ITransformerRequireStatement>;
