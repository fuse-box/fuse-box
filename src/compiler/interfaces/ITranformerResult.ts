import { ASTNode } from './AST';
import { ITransformerRequireStatementCollection } from './ITransformerRequireStatements';

export interface ITransformerResult {
  ast?: ASTNode;
  requireStatementCollection: ITransformerRequireStatementCollection;
}
