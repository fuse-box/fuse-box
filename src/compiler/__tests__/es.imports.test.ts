import { ImportType } from '../interfaces/ImportType';
import { testTranspile } from '../transpilers/testTranspiler';

/**
 to test:
    import * as hey from "./oi"
    hey.something();
*/

describe('Es imports tests', () => {
  describe('Local variable replacement', () => {
    it('should not collide with local scope', () => {
      const result = testTranspile({
        code: `
            import { foo, bar as stuff } from "./hello";
            function hello(foo) {
                console.log(foo, stuff);
            }
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should not collide with local scope 2', () => {
      const result = testTranspile({
        code: `
            import { foo, bar as stuff } from "./hello";
            function hello(foo) {
                console.log(foo, stuff);
            }
            console.log(foo, stuff)
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import something a trace it down', () => {
      const result = testTranspile({
        code: `
          import {foobar} from "foo"
          console.log(1);
          foobar();
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import something a trace it down 2', () => {
      const result = testTranspile({
        code: `
          import {FooBar} from "foo"
          console.log(1);
          new FooBar();
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import something a trace it down 3', () => {
      const result = testTranspile({
        code: `
          import {Foobar} from "foo"
          console.log(Foobar)
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import something a trace it down 4', () => {
      const result = testTranspile({
        code: `
          import {Foobar} from "foo"
          console.log([Foobar])
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import something a trace it down 5', () => {
      const result = testTranspile({
        code: `
          import FooBar from "foo"
          console.log(FooBar)
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import multiple and trace it down', () => {
      const result = testTranspile({
        code: `
          import a, {b} from "c";
          console.log(a, b)
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import multiple and trace it down 2', () => {
      const result = testTranspile({
        code: `
          import {a, b} from "c";
          console.log(a, b)
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import everything and replace in an object', () => {
      const result = testTranspile({
        code: `
          import {oi} from "c";
          const a = {
            oi : oi(String)
          }
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should import with alias', () => {
      const result = testTranspile({
        code: `
          import { ng as angular } from './angular';
          angular.module()
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import everything use it', () => {
      const result = testTranspile({
        code: `
          import * as tslib_1 from "tslib";
          tslib_1.something()
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('Should import everything use it 2', () => {
      const result = testTranspile({
        code: `
        import MySuperClass, * as everything from "module-name";
        everything.something();
        new MySuperClass();
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import everything use it 3', () => {
      const result = testTranspile({
        code: `
        import  * as everything from "module-name";
        console.log(everything)

        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import everything amd remove it (override)', () => {
      const result = testTranspile({
        code: `
        import * as everything from "module-name";
        const everything = {}
        console.log(everything)

        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should trace down assignable', () => {
      const result = testTranspile({
        code: `
          import { foo } from "foo";
          Oi.prototype[foo] = function(){}
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import and replace in extends', () => {
      const result = testTranspile({
        code: `
          import { Foo } from "foo";
          class App extends Foo {

          }
        `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('Should import with sideeffects', () => {
      const result = testTranspile({
        code: `
                  console.log(1);
                  import "foo"
                  console.log(2);
              `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should not mess with the scope 2', () => {
      const result = testTranspile({
        code: `
          import {foo, bar} from "foo"
          const foo = 1;
          console.log(foo, bar);
        `,
      });

      expect(result.code).toMatchSnapshot();
    });

    it('should not mess with the scope 3', () => {
      const result = testTranspile({
        code: `
        import func from './function.js';

        var mixin = (function(func) {
          console.log(func)
        });
        `,
      });

      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Import interfaces', () => {
    it('should ignore the import of an interface', () => {
      const result = testTranspile({
        code: `
        import { Type } from './types';

        export function hey(t: Type) {
          return 1;
        }
      `,
      });
      expect(result.code).toMatchSnapshot();
    });

    it('should  keey the statement if at least one of the imports are used', () => {
      const result = testTranspile({
        code: `
        import { Type, oi } from './types';

        export function hey(t: Type) {
          return oi();
        }
      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });

  describe('Test require statement collection', () => {
    it('should not emit a require statement ', () => {
      const result = testTranspile({
        code: `
        import { Type, oi } from './types';

        export function hey(t: Type) {

        }
      `,
      });
      expect(result.requireStatementCollection).toEqual([]);
    });

    it('should emit require statement ', () => {
      const result = testTranspile({
        code: `
        import { Type, oi } from './types';

        export function hey(t: Type) {
          console.log(oi);
        }
      `,
      });

      expect(result.requireStatementCollection).toEqual([
        {
          importType: ImportType.FROM,
          statement: {
            type: 'CallExpression',
            callee: { type: 'Identifier', name: 'require' },
            arguments: [{ type: 'Literal', value: './types' }],
          },
        },
      ]);
    });

    it('should emit raw import statement ', () => {
      const result = testTranspile({
        code: `
          import "./foo"
      `,
      });

      expect(result.requireStatementCollection).toEqual([
        {
          importType: ImportType.RAW_IMPORT,
          statement: {
            type: 'CallExpression',
            callee: { type: 'Identifier', name: 'require' },
            arguments: [{ type: 'Literal', value: './foo' }],
          },
        },
      ]);
    });
  });

  describe('Edge cases', () => {
    it('should handle lodash edge case 1', () => {
      const result = testTranspile({
        code: `
          import toString from './toString.js';
          import upperFirst from './upperFirst.js';
          function capitalize(string) {
            return upperFirst(toString(string).toLowerCase());
          }
          export default capitalize;
      `,
      });
      expect(result.code).toMatchSnapshot();
    });
  });
});
