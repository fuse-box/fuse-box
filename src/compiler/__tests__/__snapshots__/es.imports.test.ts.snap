// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Es imports tests Edge cases should handle lodash edge case 1 1`] = `
"var toString_js_1 = require(\\"./toString.js\\");
var upperFirst_js_2 = require(\\"./upperFirst.js\\");
function capitalize(string) {
  return upperFirst_js_2.default(toString_js_1.default(string).toLowerCase());
}
exports.default = capitalize;
"
`;

exports[`Es imports tests Import interfaces should  keey the statement if at least one of the imports are used 1`] = `
"var types_1 = require(\\"./types\\");
function hey(t) {
  return types_1.oi();
}
exports.hey = hey;
"
`;

exports[`Es imports tests Import interfaces should ignore the import of an interface 1`] = `
"function hey(t) {
  return 1;
}
exports.hey = hey;
"
`;

exports[`Es imports tests Local variable replacement Should import and replace in extends 1`] = `
"var foo_1 = require(\\"foo\\");
class App extends foo_1.Foo {}
"
`;

exports[`Es imports tests Local variable replacement Should import everything amd remove it (override) 1`] = `
"const everything = {};
console.log(everything);
"
`;

exports[`Es imports tests Local variable replacement Should import everything and replace in an object 1`] = `
"var c_1 = require(\\"c\\");
const a = {
  oi: c_1.oi(String)
};
"
`;

exports[`Es imports tests Local variable replacement Should import everything use it 1`] = `
"var tslib_1 = require(\\"tslib\\");
tslib_1.something();
"
`;

exports[`Es imports tests Local variable replacement Should import everything use it 2 1`] = `
"var module_name_1 = require(\\"module-name\\");
module_name_1.something();
new module_name_1.default();
"
`;

exports[`Es imports tests Local variable replacement Should import everything use it 3 1`] = `
"var module_name_1 = require(\\"module-name\\");
console.log(module_name_1);
"
`;

exports[`Es imports tests Local variable replacement Should import multiple and trace it down 1`] = `
"var c_1 = require(\\"c\\");
console.log(c_1.default, c_1.b);
"
`;

exports[`Es imports tests Local variable replacement Should import multiple and trace it down 2 1`] = `
"var c_1 = require(\\"c\\");
console.log(c_1.a, c_1.b);
"
`;

exports[`Es imports tests Local variable replacement Should import something a trace it down 1`] = `
"var foo_1 = require(\\"foo\\");
console.log(1);
foo_1.foobar();
"
`;

exports[`Es imports tests Local variable replacement Should import something a trace it down 2 1`] = `
"var foo_1 = require(\\"foo\\");
console.log(1);
new foo_1.FooBar();
"
`;

exports[`Es imports tests Local variable replacement Should import something a trace it down 3 1`] = `
"var foo_1 = require(\\"foo\\");
console.log(foo_1.Foobar);
"
`;

exports[`Es imports tests Local variable replacement Should import something a trace it down 4 1`] = `
"var foo_1 = require(\\"foo\\");
console.log([foo_1.Foobar]);
"
`;

exports[`Es imports tests Local variable replacement Should import something a trace it down 5 1`] = `
"var foo_1 = require(\\"foo\\");
console.log(foo_1.default);
"
`;

exports[`Es imports tests Local variable replacement Should import with sideeffects 1`] = `
"console.log(1);
require(\\"foo\\");
console.log(2);
"
`;

exports[`Es imports tests Local variable replacement Should trace down assignable 1`] = `
"var foo_1 = require(\\"foo\\");
Oi.prototype[foo_1.foo] = function () {};
"
`;

exports[`Es imports tests Local variable replacement should import with alias 1`] = `
"var angular_1 = require(\\"./angular\\");
angular_1.ng.module();
"
`;

exports[`Es imports tests Local variable replacement should not collide with local scope 1`] = `
"var hello_1 = require(\\"./hello\\");
function hello(foo) {
  console.log(foo, hello_1.bar);
}
"
`;

exports[`Es imports tests Local variable replacement should not collide with local scope 2 1`] = `
"var hello_1 = require(\\"./hello\\");
function hello(foo) {
  console.log(foo, hello_1.bar);
}
console.log(hello_1.foo, hello_1.bar);
"
`;

exports[`Es imports tests Local variable replacement should not mess with the scope 2 1`] = `
"var foo_1 = require(\\"foo\\");
const foo = 1;
console.log(foo, foo_1.bar);
"
`;

exports[`Es imports tests Local variable replacement should not mess with the scope 3 1`] = `
"var mixin = function (func) {
  console.log(func);
};
"
`;
